cpp/oneapi/dal/distributed/spmd_communicator.hpp

```diff
@@ -0,0 +1,153 @@
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#pragma once
+
+#include "oneapi/dal/detail/common.hpp"
+
+namespace oneapi::dal::preview {
+
+class spmd_request_iface {
+public:
+    virtual ~spmd_request_iface() = default;
+    virtual void wait() = 0;
+    virtual bool test() = 0;
+};
+
+class spmd_communicator_iface {
+public:
+    virtual ~spmd_communicator_iface() = default;
+
+    virtual std::int64_t get_rank() = 0;
+    virtual std::int64_t get_rank_count() = 0;
+
+    virtual spmd_request_iface* bcast(byte_t* send_buf, std::int64_t count, std::int64_t root) = 0;
+
+    virtual spmd_request_iface* gather(const byte_t* send_buf,
+                                       std::int64_t send_count,
+                                       byte_t* recv_buf,
+                                       std::int64_t recv_count,
+                                       std::int64_t root) = 0;
+
+    virtual spmd_request_iface* gatherv(const byte_t* send_buf,
+                                        std::int64_t send_count,
+                                        byte_t* recv_buf,
+                                        const std::int64_t* recv_count,
+                                        const std::int64_t* displs,
+                                        std::int64_t root) = 0;
+};
+
+class communication_error : public runtime_error, public std::runtime_error {
+public:
+    using std::runtime_error::runtime_error;
+
+    const char* what() const noexcept override {
+        return std::runtime_error::what();
+    }
+};
+
+class empty_spmd_request : public spmd_request_iface {
+public:
+    empty_spmd_request() = default;
+};
+
+class spmd_request : public base {
+    friend dal::detail::pimpl_accessor;
+
+public:
+    spmd_request() = default;
+
+    void wait() {
+        if (impl_) {
+            return impl_->wait();
+        }
+    }
+
+    bool test() {
+        if (impl_) {
+            return impl_->test();
+        }
+        return true;
+    }
+
+private:
+    explicit spmd_request(spmd_request_iface* impl) : impl_(impl) {}
+    dal::detail::pimpl<spmd_request_iface> impl_;
+};
+
+class spmd_communicator : public base {
+public:
+    spmd_communicator() = delete;
+
+    std::int64_t get_rank() {
+        // TODO: Handle null impl_
+        return impl_->get_rank();
+    }
+
+    std::int64_t get_rank_count() {
+        // TODO: Handle null impl_
+        return impl_->get_rank_count();
+    }
+
+    /// Broadcasts a message from the `root` rank to all other ranks
+    ///
+    /// @param send_buff
+    /// @param count
+    /// @param root
+    /// @return The object to track the progress of the operation
+    spmd_request bcast(byte_t* send_buf, std::int64_t count, std::int64_t root) {
```

Comment:
It is too MPI-like for me. Maybe it is better to add something like "const vector_class< event > &deps={}" - it is a natural way both for SYCL and oneCCL. Using requests, it is just possible to test/wait a single request but not to build a dependency graph. It might be emulated for MPI.